### PR TITLE
feat: python server

### PR DESCRIPTION
**Pull request by WinPlay02**
- every execution needs a unique id to identify calculated values
- executes compiled safe-ds code in memory (using a finder / loader)
- pipelines are executed in child processes
- communication from child processes to the main process uses a shared memory queue
- allows connections via websocket (and http)
- listens on a provided port (default 5000)
- uses flask and flask-sock for HTTP and WS
- uses gevent as WSGI implementation

### CODE REVIEW

**Comment by lars-reimann** [comment]
I have no experience with this. Maybe the [pytest-cov documentation](https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html#) helps.

**Comment by WinPlay02** [comment]
It seems to be a problem only on Windows, as the CI passes. I annotated the tests skip execution on Windows for now.
Although they work when coverage calculation does not include multiprocessing (in `.coveragerc`, see above).

**Comment by lars-reimann** [comment]
> It seems to be a problem only on Windows, as the CI passes. I annotated the tests skip execution on Windows for now.
Although they work when coverage calculation does not include multiprocessing (in .coveragerc, see above).

I can't reproduce this on a Windows machine. `pytest --cov=safeds_runner --cov-report=xml` works fine.

File: src/safeds_runner/server/main.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    set_new_websocket_target,
+    setup_pipeline_execution,
+)
+
+
+def create_flask_app(testing: bool = False) -> flask.app.App:
+    """
+    Create a flask app, that handles all requests.
+
+    :param testing Whether the app should run in a testing context
```
Comment (lars-reimann) [inline src/safeds_runner/server/main.py:30]: Docstrings should use the Numpydoc style:

* [Specification](https://numpydoc.readthedocs.io/en/latest/format.html)
* [Project setup (4. item)](https://runner.safeds.com/en/stable/development/environment/#project-setup)
* [Project guidelines](https://library.safeds.com/en/stable/development/project_guidelines/#docstrings) (other repo, will be copied here too)

File: src/safeds_runner/server/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+"""Infrastructure for dynamically running Safe-DS pipelines and communication with the vscode extension."""
```
Comment (lars-reimann) [inline src/safeds_runner/server/__init__.py:1]:
```suggestion
"""Infrastructure for dynamically running Safe-DS pipelines and communication with the VS Code extension."""
```

**Review by lars-reimann — CHANGES_REQUESTED** [review]

**Comment by WinPlay02** [comment]
I think I addressed all the suggestions. Now, the `PipelineManager` class takes the place of most of the global variables. It is created at startup and used in the websocket session.
But I couldn't get rid of the `current_pipeline` global in `pipeline_manager.py` as it is used for saving placeholders.

File: src/safeds_runner/server/pipeline_manager.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        After that a message queue handling thread is started in the main process.
+        This allows receiving messages directly from one of the pipeline processes and relaying information
+        directly to the websocket connection (to the VS Code extension).
+        """
+        self.multiprocessing_manager: SyncManager = multiprocessing.Manager()
+        self.placeholder_map: dict = {}
+        self.messages_queue: queue.Queue[Message] = self.multiprocessing_manager.Queue()
+        self.websocket_target: simple_websocket.Server | None = None
+        self.messages_queue_thread: threading.Thread = threading.Thread(target=self._handle_queue_messages, daemon=True)
+        self.messages_queue_thread.start()
```
Comment (lars-reimann) [inline src/safeds_runner/server/pipeline_manager.py:53]: Add leading `_` to the names of attributes that should be internal to the class.

File: src/safeds_runner/server/pipeline_manager.py [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+        messages_queue : queue.Queue[Message]
+            A queue to write outgoing messages to.
+        placeholder_map : dict[str, Any]
+            A map to save calculated placeholders in.
+        """
+        self.pipeline = pipeline
+        self.id = execution_id
+        self.messages_queue = messages_queue
+        self.placeholder_map = placeholder_map
+        self.process = multiprocessing.Process(target=self._execute, daemon=True)
```
Comment (lars-reimann) [inline src/safeds_runner/server/pipeline_manager.py:157]: Add leading `_` to the names of attributes that should be internal to the class.

File: src/safeds_runner/server/pipeline_manager.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+        return "Int"
+    if isinstance(value, str):
+        return "String"
+    if isinstance(value, object):
+        object_name = type(value).__name__
+        if object_name == "function":
+            return "Callable"
+        if object_name == "NoneType":
+            return "Null"
+        return object_name
```
Comment (lars-reimann) [inline src/safeds_runner/server/pipeline_manager.py:288]: Can also be expressed using [pattern matching](https://stackoverflow.com/a/67524642). It's up to you whether you want to change it or leave it.

**Review by lars-reimann — CHANGES_REQUESTED** [review]
Looks great overall. I just have some minor requests.

**Review by lars-reimann — APPROVED** [review]
Great contribution, thanks! 👍

**Pull request merged**